File: Java_Fundamentals_January_2017/Java_Advanced/Exersices/02. Matrices C#/Matrices/Evaluate Expression/EvaluateExpression.cs
namespace Evaluate_Expression
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    public class EvaluateExpression
    {
        public static void Main(string[] args)
        {
            string[] tokens = Console.ReadLine().Split(' ');
            Queue<string> output = PolishReverse(tokens);

            decimal result = Evaluate(output);

            Console.WriteLine($"{result:f2}");

        }

        private static decimal Evaluate(Queue<string> output)
        {

            Stack<decimal> operands = new Stack<decimal>();
            Queue<string> operators = new Queue<string>();

            if (output.Count != 0)
            {
                while (output.Count != 0)
                {
                    if (IsNumber(output.Peek()))
                    {
                        operands.Push(decimal.Parse(output.Dequeue()));
                    }
                    else
                    {
                        operators.Enqueue(output.Dequeue());
                    }
                }
            }

            while (operands.Count > 1)
            {
                string op = operators.Dequeue();
                operands = CalcResult(operands, op);
            }

            return operands.Pop();
        }

        private static Stack<decimal> CalcResult(Stack<decimal> operands, string op)
        {
            decimal op2 = operands.Pop();
            decimal op1 = operands.Pop();
            switch (op)
            {
                case "+": operands.Push(op1 + op2); break;
                case "-": operands.Push(op1 - op2); break;
                case "*": operands.Push(op1 * op2); break;
                case "/": operands.Push(op1 / op2); break;
                case "^":
                    operands.Push((decimal)(Math.Pow((double)op1, (int)op2))); break;
            }

            return operands;
        }

        private static Queue<string> PolishReverse(string[] tokens)
        {
            Queue<String> output = new Queue<string>();
            Stack<string> operators = new Stack<string>();

            foreach (string token in tokens)
            {
                if (IsNumber(token))
                {
                    output.Enqueue(token);
                }
                else if (token == "(")
                {
                    operators.Push(token);
                    continue;
                }
                else if (IsRightAssociative(token))
                {
                    output.Enqueue(token);
                }
                else if (IsLeftAssociative(token))
                {
                    if (operators.Count != 0)
                    {
                        while (operators.Count > 0)
                        {
                            string op2 = operators.Peek();
                            if ((token == "+" || token == "-") && (op2 == "*" || op2 == "/")
                                || ((token == "*" || token == "/") && (op2 == "*" || op2 == "/" || op2 == "^")))
                            {
                                output.Enqueue(operators.Pop());
                            }
                            else break;
                        }
                    }

                    operators.Push(token);
                }
                else if (token == ")")
                {
                    while (operators.Peek() != "(")
                    {
                        output.Enqueue(operators.Pop());
                    }
                    operators.Pop();
                }
            }

            while (operators.Count != 0)
            {
                output.Enqueue(operators.Pop());
            }

            return output;
        }

        private static bool IsLeftAssociative(string token)
        {
            return token == "+" || token == "-" || token == "*" || token == "/";
        }

        private static bool IsRightAssociative(string token)
        {
            return token == "^";
        }

        private static bool IsNumber(string token)
        {
            string[] operators = { "+", "-", "*", "/", "(", ")" };
            return !operators.Contains(token);
        }
    }
}

File: Java_Fundamentals_January_2017/Java_Advanced/Exersices/03. Stacks/src/EvaluateExpression.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.DoubleSummaryStatistics;
import java.util.Queue;

public class EvaluateExpression {
    public static void main(String[] args) throws IOException {
        BufferedReader in =
                new BufferedReader(new InputStreamReader(System.in));

        String[] tokens = in.readLine().split("\\s+");

        Queue<String> output = PolishReverse(tokens);

        Double result = Evaluate(output);

        System.out.printf("%.2f", result);
    }

    private static Double Evaluate(Queue<String> output) {
        Deque<Double> result = new ArrayDeque<>();

        while(!output.isEmpty()){
            if(IsNumber(output.peek())){
                result.push(Double.parseDouble(output.poll()));
            }
            else{
                String op = output.poll();

                double op2 = result.pop();
                double op1 = result.pop();

                switch (op){
                    case "+": result.push(op1 + op2); break;
                    case "-": result.push(op1 - op2); break;
                    case "*": result.push(op1 * op2); break;
                    case "/": result.push(op1 / op2); break;
                }
            }
        }

        return result.pop();
    }

    private static ArrayDeque<String> PolishReverse(String[] tokens) {
        ArrayDeque<String> output = new ArrayDeque<>();
        ArrayDeque<String> operators = new ArrayDeque<>();

        for(String token : tokens)
        {
            if (IsNumber(token))
            {
                output.offer(token);
            }
            else if(token.equals("("))
            {
                operators.push(token);
            }
            else if (IsRightAssociative(token))
            {
                output.offer(token);
            }
            else if (IsLeftAssociative(token))
            {
                if (!operators.isEmpty())
                {
                    while (!operators.isEmpty()){
                        String op2 = operators.peek();
                        if (((token.equals("+") || token.equals("-"))
                                && IsLeftAssociative(op2))
                                || ((token.equals("*") || token.equals("/"))
                                && (op2.equals("*") || op2.equals("/") || op2.equals("^"))))
                        {
                            output.offer(operators.pop());
                        }
                        else break;
                    }
                }

                operators.push(token);
            }
            else if (token.equals(")"))
            {
                while (!operators.peek().equals("("))
                {
                    output.offer(operators.pop());
                }
                operators.pop();
            }
        }

        while (!operators.isEmpty())
        {
            output.offer(operators.pop());
        }

        return output;
    }

    private static boolean IsLeftAssociative(String token)
    {
        return token.equals("+") || token.equals("-") || token.equals("*") || token.equals("/");
    }

    private static boolean IsRightAssociative(String token)
    {
        return token.equals("^");
    }

    private static boolean IsNumber(String token)
    {
        String[] ops = {"+", "-", "*", "/", "(", ")"};
        for(String op  : ops){
            if(token.equals(op)) return false;
        }

        return true;
    }
}
